apps/src/dance/DanceAiModal.tsx

```diff
@@ -0,0 +1,321 @@
+import React, {useState} from 'react';
+import moduleStyles from './dance-ai.module.scss';
+import AccessibleDialog from '@cdo/apps/templates/AccessibleDialog';
+import Button from '@cdo/apps/templates/Button';
+import HttpClient from '@cdo/apps/util/HttpClient';
+import {CHAT_COMPLETION_URL} from '@cdo/apps/aichat/constants';
+import {useSelector} from 'react-redux';
+import {useAppDispatch} from '@cdo/apps/util/reduxHooks';
+import {setShowingAi} from './danceRedux';
+import {StrongText, Heading5} from '@cdo/apps/componentLibrary/typography';
+import classNames from 'classnames';
+import FontAwesome from '@cdo/apps/templates/FontAwesome';
+const Typist = require('react-typist').default;
+
+const aiBotBorder = require('@cdo/static/dance/ai-bot-border.png');
+
+const promptString = 'Generate a scene using this mood:';
+
+const doAi = async (input: string) => {
+  const systemPrompt =
+    'You are a helper which can accept a request for a mood or atmosphere, and you then generate JSON like the following format: {backgroundColor: "black", backgroundEffect: "splatter", foregroundEffect: "rain"}.  The only valid values for backgroundEffect are circles, color_cycle, diamonds, disco_ball, fireworks, swirl, kaleidoscope, lasers, splatter, rainbow, snowflakes, galaxy, sparkles, spiral, disco, stars.  The only valid values for backgroundColor are rave, cool, electronic, iceCream, neon, tropical, vintage, warm.  The only valid values for foregroundEffect are bubbles, confetti, hearts_red, music_notes, pineapples, pizzas, smiling_poop, rain, floating_rainbows, smile_face, spotlight, color_lights, raining_tacos.  Make sure you always generate all three of those values.  Also, add a field called "explanation" to the result JSON, which contains a simple explanation of why you chose the values that you did, at the reading level of a 5th-grade school student, in one sentence of less than forty words.';
+
+  const messages = [
+    {
+      role: 'system',
+      content: systemPrompt,
+    },
+    {
+      role: 'user',
+      content: input,
+    },
+  ];
+
+  console.log('do AI:', input);
+
+  const response = await HttpClient.post(
+    CHAT_COMPLETION_URL,
+    JSON.stringify({messages}),
+    true,
+    {
+      'Content-Type': 'application/json; charset=UTF-8',
+    }
+  );
+
+  if (response.status === 200) {
+    const res = await response.json();
+    return res.content;
+  } else {
+    return null;
+  }
+};
+
+interface Item {
```

Comment:
Done.  

(Given that they were very specific to the modal file, I'm freshly curious about the benefits of having a common types file, if you or @sanchitmalhotra126 have thoughts.)

Reply:
Thanks! I'm very new to TS and learning loads from @sanchitmalhotra126 but it does seem nice to set a pattern and also shortens the react file making the react code easier to follow./read.

Reply:
Seems good to me! Deciding where to put types and interfaces can be a bit vague, and I think we're still developing some norms around this as an engineering team. My general instinct is that if a type is going to be used in more than one place, then it belongs in a separate file, whether that's a `types.ts` or its own `SomeInterface.ts`. In contrast, I think React prop types are a good example of keeping types inside the same file, since they're only referenced directly by the react component.